### PR TITLE
Further abstraction of the dependency system for non-cube dependents and bugfixing of the dependency system

### PR DESCRIPTION
**Pull request by Cyclonit**
The 2d visualization code you posted does not show any unloading of cubes. However, I did make the ServerCubeCache log whenever it unloaded a cube (and loaded) and it did appear to act properly. I couldn't find any cubes not being unloaded after moving 1000 blocks to the north. One issue remaining is the world not showing up on the first generation. Saving the game and reloading solves the problem.

### CODE REVIEW

File: src/main/java/cubicchunks/server/ServerCubeCache.java [hunk 1]
```diff
@@ -24,6 +24,8 @@
 package cubicchunks.server;
 
 import com.google.common.collect.Maps;
+
+import cubicchunks.CubeProviderDebug;
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/server/ServerCubeCache.java]: Make sure to remove this before commit, it breaks the build

File: src/main/java/cubicchunks/server/ServerCubeCache.java [hunk 2]
```diff
@@ -104,12 +106,7 @@ public ServerCubeCache(ICubicWorldServer worldServer) {
 		this.forceAdded = new HashMap<>();		
 		this.forceAddedReverse = new HashMap<>();
 	}
-
-	@Override
-	public Collection<Chunk> getLoadedChunks() {
-		return (Collection<Chunk>) (Object) this.loadedColumns.values();
-	}
-
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/server/ServerCubeCache.java]: I think this is required

File: src/main/java/cubicchunks/worldgen/GeneratorPipeline.java [hunk 3]
```diff
@@ -102,10 +102,6 @@ public void generate(Cube cube, GeneratorStage targetStage) {
 			if (dependency != null) {
 				Dependent dependent = new Dependent(cube, dependency);
 				this.dependencyManager.register(dependent);
-
-				if (dependent.isSatisfied()) {
-					stage.getProcessor().processor.add(cube.getAddress());
-				}
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/GeneratorPipeline.java]: wait, what? How does it even work?

File: src/main/java/cubicchunks/worldgen/dependency/DependencyManager.java [hunk 4]
```diff
@@ -137,7 +130,7 @@ private void addNewRequirement(Dependent dependent, Requirement requirement) {
 		
 		// If the cube is loaded, update the dependent.
 		if (requiredCube != null) {
-			dependent.update(this, requiredCube);
+//			dependent.update(this, requiredCube);
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/dependency/DependencyManager.java]: Don't comment out code. That's what vercion control is for. Why is it removed?

File: src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java [hunk 5]
```diff
@@ -36,40 +36,24 @@
...
-		
-		this.xLow = -radius;
-		this.xHigh = radius;
-		this.yLow = -radius;
-		this.yHigh = radius;
-		this.zLow = -radius;
-		this.zHigh = radius;
-	}
-	
-	public RegionDependency(Cube cube, GeneratorStage stage, int xLow, int xHigh, int yLow, int yHigh, int zLow, int zHigh) {
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java]: This constructor was really useful. It was wrong but after some changes very useful. Maybe readd it with 2 `Vec3i`s as arguments?

File: src/main/java/cubicchunks/worldgen/GeneratorPipeline.java (hunk 3 above)
Comment (Cyclonit) [inline src/main/java/cubicchunks/worldgen/GeneratorPipeline.java]: This is not required due to DependencyManager.register(...) queuing the cube if all dependencies are satisfied.

File: src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java (hunk 5 above)
Comment (Cyclonit) [inline src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java]: Which way do you prefer the Dependencies to work? One instance for all cubes or individual instances for each? This constructor was a relict from when I used the latter.

Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java]: One instance for all cubes is definitely better. I just want to be able to specify region that isn't a cube, which is what this kind of constructor would be useful for.

File: src/main/java/cubicchunks/server/CubeIO.java [hunk 6]
```diff
@@ -144,7 +144,8 @@ public void run() {
 	public void flush() {
 		if (columnsToSave.size() != 0 || cubesToSave.size() != 0) {
 			err("Attempt to flush() CubeIO when there are remaining cubes to save! Saving remaining cubes to avoid corruption");
-			while (this.writeNextIO()) ;
+			while (this.writeNextIO())
+				;
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/server/CubeIO.java]: ...

File: src/main/java/cubicchunks/server/CubeIO.java [hunk 7]
```diff
@@ -285,9 +286,8 @@ public boolean writeNextIO() {
 
 			long diff = System.currentTimeMillis() - start;
 			LOGGER.debug("Wrote {} columns ({} remaining) ({}k) and {} cubes ({} remaining) ({}k) in {} ms",
-					numColumnsSaved, numColumnsRemaining, numColumnBytesSaved/1024,
-					numCubesSaved, numCubesRemaining, numCubeBytesSaved/1024, diff
-			);
+					numColumnsSaved, numColumnsRemaining, numColumnBytesSaved / 1024,
+					numCubesSaved, numCubesRemaining, numCubeBytesSaved / 1024, diff);
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/server/CubeIO.java]: seriously, try to avoid reformatting stuff.

**Comment by Barteks2x** [comment]
Which world type did you test it with? I think I've seen some columns being left loaded as I'm moving, see #11.

Also, this issue needs fixing: #14

**Comment by Cyclonit** [comment]
I tested using Custom Cubic.

**Comment by Barteks2x** [comment]
Just squash all of these small commits... this is getting a bit too messy ([this](https://robots.thoughtbot.com/git-interactive-rebase-squash-amend-rewriting-history) may be useful to read if you don't know what it means)

**Comment by Cyclonit** [comment]
I'll clean it up asap. I managed to confuse origin/upstream and master/dependency, leading to a jambled mess... Sorry

File: src/main/java/cubicchunks/server/CubeIO.java [hunk 8]
```diff
@@ -95,7 +95,7 @@ private static DB initializeDBConnection(final File saveFile, final WorldProvide
 
 		file.getParentFile().mkdirs();
 
-		DB db = DBMaker.fileDB(file).transactionEnable().make();
+		DB db = DBMaker.fileDB(file).make();
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/server/CubeIO.java]: transactions should be enabled

File: src/main/java/cubicchunks/server/CubeIO.java [hunk 9]
```diff
@@ -542,7 +542,7 @@ private static NBTTagCompound writeCubeToNbt(final Cube cube) {
 		// tile entities
 		NBTTagList nbtTileEntities = new NBTTagList();
 		nbt.setTag("TileEntities", nbtTileEntities);
-		for (TileEntity blockEntity : cube.getTileEntityMap().values()) {
+		for (TileEntity blockEntity : cube.getTileEntityMap()) {
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/server/CubeIO.java]: This is just wrong. cube.getTileEntityMap() returns a Map, which isn't Iterable.

Comment (Cyclonit) [inline src/main/java/cubicchunks/server/CubeIO.java]: I wholeheartedly agree and I honestly have no idea why that changed. I do not recall changing that line or anything in CubeIO for that matter. Honestly, I should not have started working on this today, I'm sorry for causing this much work.

File: src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java [hunk 10]
```diff
@@ -36,40 +36,24 @@
...
-	private int xLow;
-	private int xHigh;
-	private int yLow;
-	private int yHigh;
-	private int zLow;
-	private int zHigh;
 	
-	
-	public RegionDependency(GeneratorStage targetStage, int radius) {
+	public RegionDependency(Cube cube, GeneratorStage targetStage, int radius) {
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java]: So is dependency per cube or one for all cubes?

Comment (Cyclonit) [inline src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java]: Currently its per cube. I considered it to be somewhat more clean, because this permits stateaware dependencies per cube. But if you prefer singletons, I can change it.

Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java]: I would make it something like asking DependencyProvider for a dependency for each cube separately and let the DependencyProvider decide if it wants to use singleton or provide different dependency for each cube (and remove the Cube argument from constructor).

Comment (Cyclonit) [inline src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java]: I'll continue working on this tomorrow... Right now I'm >this< close to punching myself in the face. I have no idea how I managed to break my repository this bad.

**Comment by Barteks2x** [comment]
I would also really suggest switching to IntelliJ IDEA, at least for this project. This way you will avois almost all formatting changes.

File: src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java (hunk 10 above)
Comment (Cyclonit) [inline src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java]: The proposed flexibility is already there. The GeneratorPipeline calls `GeneratorStage.getDependency(cube) (== DependencyProvider.getDependency(cube)`
It does not matter if the stage uses a singleton or an individual instance per cube, because the calls to Dependency.getRequirements and DependencyManager.update receive the cube as parameter. I've changed RegionDependency to act as a Singleton as you suggested.

**Comment by Cyclonit** [comment]
I have changed to IDEA, but I cannot guarantee an immediate improvement with regards to formatting as it is driving me nuts D:

**Comment by Barteks2x** [comment]
Just import my formatter settings. This way you will have exactly the same settings.

File: src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java [hunk 11]
```diff
@@ -26,75 +26,73 @@
...
+		this.xLow = -radius;
+		this.xHigh = radius;
+		this.yLow = -radius;
+		this.yHigh = radius;
+		this.zLow = -radius;
+		this.zHigh = radius;
 	}
 	
-	public RegionDependency(Cube cube, GeneratorStage stage, int xLow, int xHigh, int yLow, int yHigh, int zLow, int zHigh) {
+	public RegionDependency(GeneratorStage targetStage, int xLow, int xHigh, int yLow, int yHigh, int zLow, int zHigh) {
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java]: I would suggest using `Vec3i`s instead of ints here, argument order isn't obvious from code.

Comment (Cyclonit) [inline src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java]: Do you have any naming in mind? "lowerFrontLeft" and "upperBackRight" sound kinda wrong...

Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java]: startPosition and endPosition? You can sort them in constructor.

File: src/main/java/cubicchunks/worldgen/dependency/DependencyManager.java [hunk 12]
```diff
@@ -36,139 +36,92 @@
 import java.util.Map;
 import java.util.Set;
 
+/*
+ * TODO: Abstraction to allow for non-cube dependents.
+ * TODO: Commenting
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/dependency/DependencyManager.java]: Is it going to be done in this PR?

Comment (Cyclonit) [inline src/main/java/cubicchunks/worldgen/dependency/DependencyManager.java]: The commenting definitely, the non-cube dependents depending on how soon you'd like to merge.

Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/dependency/DependencyManager.java]: Non-cube dependents are not needed right now, so it can wait. Is issue #14 going to be fixed? This is a critical issue for the dependency system.

Comment (Cyclonit) [inline src/main/java/cubicchunks/worldgen/dependency/DependencyManager.java]: I'm working on it.

Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/dependency/DependencyManager.java]: Just a note as you may not know it, when you fix it, add `Fixes #14` in your commit message to automatically close the issue when it's merged.

**Comment by Cyclonit** [comment]
I found one bug causing dependencies of live cube's not being unregistered causing lingering cubes. I went ahead and placed a simple log-output in unloadCube
`		if (dependencyManager.isRequired(new CubeCoords(cubeX, cubeY, cubeZ))) {
			CubicChunks.LOGGER.info("Not unloading cube at: ({},{},{})", cubeX, cubeY, cubeZ);
			return;
		}`
This line is never called and I don't understand why. As soon as I move away from a region currently being generated, this line should come up, because the furthest away cubes should be kept alive by those closer to the player. Sadly, that's all I have time for today.

File: src/main/java/cubicchunks/worldgen/dependency/DependencyManager.java [hunk 13]
```diff
@@ -99,16 +97,23 @@ private void addNewRequirement(DependentCube dependentCube, Requirement requirem
...
-		// Remember the dependentCube.
-		CubeCoords coords = dependentCube.getCube().getCoords();
-		this.dependentMap.put(coords, dependentCube);
+		// Get the requirements
+		Collection<Requirement> requirements = dependentCube.getRequirements();
 
-		for (Requirement requirement : dependentCube.getRequirements()) {
-			addNewRequirement(dependentCube, requirement);
-		}
+		if (requirements.size() > 0) {
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/dependency/DependencyManager.java]: why is this needed?

File: src/main/java/cubicchunks/worldgen/dependency/DependencyManager.java [hunk 14]
```diff
@@ -30,156 +30,113 @@
...
-		
-		for (Requirement requirement : dependent.getRequirements()) {
-			addNewRequirement(dependent, requirement);
-		}
+	public void register(DependentCube dependentCube) {
+
+		// Get the requirements
+		Collection<Requirement> requirements = dependentCube.getRequirements();
+
+		if (requirements.size() > 0) {
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/dependency/DependencyManager.java]: Why is this required?

**Comment by Barteks2x** [comment]
Still waiting for updates on it, there is a critical bug that needs to be fixed. Right now the dependency system is in unusable state, so switching to the new system is impossible.

I have no idea how you are testing it that it works for you, I will add some code that uses the dependency system to CubicDebug world type.

**Comment by Cyclonit** [comment]
I'll still do some testing to be sure, but thus far all required cubes are being loaded and are available when needed. However, there is still some weirdness going in terms of some requirements not being cleared leading to the list of required cubes having rogue entries.

File: src/main/java/cubicchunks/world/ICubeCache.java [hunk 15]
```diff
@@ -23,14 +23,19 @@
...
 package cubicchunks.world;
 
+import cubicchunks.util.CubeCoords;
 import cubicchunks.world.column.Column;
 import cubicchunks.world.cube.Cube;
 
 public interface ICubeCache {
 	boolean cubeExists(int cubeX, int cubeY, int cubeZ);
 
+	boolean cubeExists(CubeCoords coords);
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/world/ICubeCache.java]: this can be default method with `return this.getCube(coords) == null`. Client should contain all cubes, and return BlankCube for cubes that don't exist.

Comment (Cyclonit) [inline src/main/java/cubicchunks/world/ICubeCache.java]: I don't understand quite what you mean. In both ServerCubeCache and ClientCubeCache this method acts as a proxy for cubeExists(x, y, z). Therefor on the client it always returns true as required.

File: src/main/java/cubicchunks/world/dependency/DependencyManager.java [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+			if (requiredCube == null) {
+
+				// If the required cube is loaded already, add it to the RequiredCube instance immediately.
+				Cube cube = this.cubeProvider.getCube(requirement.getCoords());
+				if (cube != null) {
+					requiredCube = new RequiredCube(cube);
+				}
+				// Otherwise create an empty instance.
+				else {
+					requiredCube = new RequiredCube();
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/world/dependency/DependencyManager.java]: Why these 2 need to be separate? Why `new RequiredCube(null)` can't be equivalent to `new RequiredCube()`?

Comment (Cyclonit) [inline src/main/java/cubicchunks/world/dependency/DependencyManager.java]: They are equivalent, its personal taste of mine to keep things like this separate as it increases readability (for me). A parameter being null just isn't the same as it not being there. If you want me to change this I'll do so.

File: src/main/java/cubicchunks/world/dependency/DependencyManager.java [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+
+			// Add the dependent to the required cubes list of dependents.
+			requiredCube.addDependent(dependent);
+
+			// If the required cube is loaded, notify the dependent.
+			if (requiredCube.isAvailable()) {
+				dependent.update(this, requiredCube.getCube());
+
+				// Has the cube reached its target stage?
+				if (requiredCube.getCube().getCurrentStage() == requiredCube.getCube().getTargetStage()) {
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/world/dependency/DependencyManager.java]: What if the cube is after the target stage? I would suggest using `!requiredCube.getCube().getCurrentStage().precedes(requiredCube.getCube().getTargetStage())` (or, with my recent changes, `!requiredCube.getCube.isBeforeStage(requiredCube.getCube.getTargetStage)`)

Comment (Cyclonit) [inline src/main/java/cubicchunks/world/dependency/DependencyManager.java]: As of right now it cannot be after its target stage unless something went wrong. The only component advancing stages is the GeneratorPipeline, which does not advance past the target stage. However, using your method is cleaner and I will change it.

File: src/main/java/cubicchunks/world/dependency/DependencyManager.java [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+	 * {@link #isRequired(CubeCoords coords)}
+	 *
+	 * @return True iff there exists at least one dependent requiring the given cube to remain loaded.
+	 */
+	public boolean isRequired(Cube cube) {
+		return this.isRequired(cube.getCoords());
+	}
+
+	public void tick() {
+		CubicChunks.LOGGER.info("{}", this.requiredMap.size());
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/world/dependency/DependencyManager.java]: Does it really run every tick? If so, I would suggest changing log level to DEBUG.

Comment (Cyclonit) [inline src/main/java/cubicchunks/world/dependency/DependencyManager.java]: This should not have made it into the commit. It will be removed.

File: src/main/java/cubicchunks/world/dependency/RequiredCube.java [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+
+
+	public RequiredCube() {
+		this.cube = null;
+		this.dependents = new HashSet<>();
+	}
+
+	public RequiredCube(Cube cube) {
+		this.cube = cube;
+		this.dependents = new HashSet<>();
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/world/dependency/RequiredCube.java]: Why  use 2 constructors at all of `new RequiredCube(null)` is equivalent to `new RequiredCube()`?

Comment (Cyclonit) [inline src/main/java/cubicchunks/world/dependency/RequiredCube.java]: A RequiredCube without its cube property being set represents a cube which is required but not loaded. To me, this is important to point out. Therefor I decided to go with this layout, to make the difference apparent to other people looking at the code. As I said on your earlier comment, I am willing to remove this if you want me to. If you don't, I'd add comments to bring the point home a little better.

File: src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java [hunk 20]
```diff
@@ -77,7 +81,7 @@ public RegionDependency(GeneratorStage targetStage, Vec3i relA, Vec3i relB) {
 	}
 
 	public boolean update(DependencyManager manager, DependentCube dependentCube, Cube requiredCube) {
-		return requiredCube.getCurrentStage() == targetStage;
+		return !requiredCube.getCurrentStage().precedes(targetStage);
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/dependency/RegionDependency.java]: Name of this method doesn't really say what it actually does.

File: src/main/java/cubicchunks/world/ICubeCache.java (hunk 15 above)
Comment (Barteks2x) [inline src/main/java/cubicchunks/world/ICubeCache.java]: right, ok. Just leave it then. Somehow I missed their implementations.

File: src/main/java/cubicchunks/world/dependency/DependencyManager.java (hunk 16 above)
Comment (Barteks2x) [inline src/main/java/cubicchunks/world/dependency/DependencyManager.java]: Ok, but then you probably should use `@Nullable`/`@NotNull` to specify it (look at `ICubicWorld` to see how they are used)

**Comment by Barteks2x** [comment]
Can you update PR title to say what is actually changed?

**Comment by Cyclonit** [comment]
Well... I just spent 2 hours trying to figure out why certain Dependents did not end up being processed. Turns out I forgot to add a Dependency for the feature generator leading to the bottom most layer of cubes not moving past the feature state.

File: src/main/java/cubicchunks/world/dependency/Dependency.java [hunk 21]
```diff
@@ -26,31 +26,32 @@
...
 
 	/**
-	 * This collection specifies which cubes must be loaded for a given cube's requirements to be satisfied.
+	 * Provides a collection of requirements specifying the given cube's required cubes and the GeneratorStages they
+	 * need to be at.
 	 *
 	 * @return A collection of Requirements specifying the given cube's requirements.
 	 */
-	Collection<Requirement> getRequirements(Cube cube);
+	@Nonnull
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/world/dependency/Dependency.java]: You can add `@MethodsReturnNonnullByDefault` and `@ParametersAreNonnullByDefault` annotations on the whole class to make methods `@Nonnull` by default, and then only add `@Nullable` when needed.

File: src/main/java/cubicchunks/worldgen/generator/custom/CustomFeatureProcessor.java [hunk 22]
```diff
@@ -67,9 +67,6 @@ public CustomFeatureProcessor(GeneratorStage generatorStage, String name, ICubeC
 	@Override
 	public Set<Cube> calculate(Cube cube) {
 
-		if (!this.canGenerate(cube)) {
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/generator/custom/CustomFeatureProcessor.java]: I'm almost sure that after removing this, the canGenerate method is unused and can be removed.

**Comment by Barteks2x** [comment]
Note: Commit cd96f087b65130ecfe084498a608998d0d924fd2 (and 75621f8b713a55771a424d74761d7e938b956c8a) may have some bad issues with chunk loading. Don't merge them until I fix it, it will cause a lot of weird issues.

**Comment by Barteks2x** [comment]
Aside of what I mentioned in comment to the issue, here are some of my observations (it's after merging it into master):
* Lighting is more broken then it was before - random areas are fully lit, and in some areas lighting isn't diffused properly
* Chunks are always generated instead of loaded from disk

**Comment by Cyclonit** [comment]
Fixed no2.

**Comment by Cyclonit** [comment]
I've added a check to FirstLightingProcessor.calculate to determine if all required cubes were in fact loaded and in 10 test runs I did not encounter a single missing cube.

**Comment by Barteks2x** [comment]
Did you also test if they are correct stage? Did you test that cube stages are not being calculated out of order (it was the root cause of my last lighting bug)?

If it's neighter of these, I can figure it out myself. Last time I debugged `FirstLightProcessor` I ended up with 130MB log file, hopefully it will be simplier this time.

**Comment by Cyclonit** [comment]
Well... turns out I'm stupid. DependentCube decremented a counter whenever a cube matching the RegionDependency's requirements was updated. This caused cubes moving from lighting->population and from population->live to decrements the counter twice, therefor causing DependentCube to release too soon. Race conditions like this are fun! Fixed by making Requirement stateful and remembering which Requirement was fulfilled.

However, this still doesn't fix the rogue cube issue.

**Comment by Barteks2x** [comment]
I think you could reuse my load cause tracking code from `DebugCubeProvider` (I added it in latest revision) to track from where and how cube stage got updated last time.

File: src/main/java/cubicchunks/lighting/FirstLightProcessor.java [hunk 23]
```diff
@@ -74,6 +75,18 @@ public FirstLightProcessor(GeneratorStage lighting, String name, ICubeCache cach
 		ICubicWorld world = cube.getWorld();
 		ICubeCache cache = world.getCubeCache();
 
+		for (int x = -2; x <= 2; ++x) {
+			for (int y = -2; y <= 2; ++y) {
+				for (int z = -2; z <= 2; ++z) {
+					CubeCoords coords = new CubeCoords(cube.getX() + x, cube.getY() + y, cube.getZ() + z);
+					if (!cache.cubeExists(coords) || cache.getCube(coords).getCurrentStage().precedes(generatorStage)) {
+						CubicChunks.LOGGER.error("Missing cube ({})! {}/{}", coords, cache.getCube(coords).getCurrentStage().getName(), generatorStage.getName());
+						System.exit(1);
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/lighting/FirstLightProcessor.java]: 1. NEVER use System.exit(). Use `FMLCommonHandler.exitJava()`. Or throw `new Error().`
2. You can simply create `new  FastCubeBlockAccess(cubeCache, centerCube, radius)` and modify it to also check stage - it will throw exception if required cube isn't loaded and it will dump information about which cubes are loaded (you can also modify that to include stage if it doesn't do that now)

Comment (Cyclonit) [inline src/main/java/cubicchunks/lighting/FirstLightProcessor.java]: I just quickly commited my current code so you can have a go at it if you want to. The check in FirstLightProcessor can go as it never encounters a missing block. If the DependencySystem were failing at delivering the required cubes properly, that loop would exit the program and it does not.

Comment (Barteks2x) [inline src/main/java/cubicchunks/lighting/FirstLightProcessor.java]: Ok, just keep in mind that before merging it, `System.exit()` must be removed. Or possibly remove this check entirely, and then check that `world.checkLightFor()` returns `true`.

Comment (Cyclonit) [inline src/main/java/cubicchunks/lighting/FirstLightProcessor.java]: Of course, I'll clean it up once I found this f*** bug...

**Comment by Barteks2x** [comment]
I'm not sure what the issue is now. I'm flying around for several minutes and din't experience any bug. Lighting works as I expect it to work and it shows no rogue cubes.

**Comment by Cyclonit** [comment]
Try logging the DependentCubeManager.getDependentCubeCount and DependencyManager.getRequiredCubeCount. Those should approach zero, but sometimes they don't.

**Comment by Barteks2x** [comment]
As for these cubes that are not sent to player: they seem to have wrong target stage set. After trying to see why `PlayerCubeMap` doesn't send it to me it turned out that target stage is set to lighting. And I was in that cube.

**Comment by Barteks2x** [comment]
Also, as of some recent changes FeatureProcessor has no dependencies as I merged surface and terrain stages. And (at least in my case) debugging showed that it's this stage that has some issues (still, it would be better to fix it properly if it works with Features as independent stage)

**Comment by Cyclonit** [comment]
I'm somewhat confident now that everything works as it was supposed to do last week...

Thanks for pointing out the wrong targetStage. Apparently some reorganization of loadCube caused it not being set when loadCube was invoked with a targetStage as parameter.

File: src/main/java/cubicchunks/worldgen/GeneratorPipeline.java [hunk 24]
```diff
@@ -122,7 +122,7 @@ public void generate(Cube cube) {
 	public void generate(Cube cube, GeneratorStage targetStage) {
 
 		// Make sure the proper target stage is set.
-		if (cube.getTargetStage().precedes(targetStage)) {
+		if (cube.getTargetStage() == null || cube.getTargetStage().precedes(targetStage)) {
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/worldgen/GeneratorPipeline.java]: Can target stage be null?

Comment (Cyclonit) [inline src/main/java/cubicchunks/worldgen/GeneratorPipeline.java]: Currently yes, because it is no longer set in all paths of ServerCubeCache.loadCube, but I'll change that to get rid of that additional check. Cleaner without it.

**Comment by Barteks2x** [comment]
Chunk unloading still needs to be fixed, this PR breaks it:
unloading

**Comment by Cyclonit** [comment]
The dependency system makes it difficult for cubes to be unloaded during generation. Usually they end up being finished and unloaded afterwards, meaning it may take some time for them to disappear.

**Comment by Barteks2x** [comment]
The way it works is a bit weird. It works fine when moving vertically, but when moving in x/z - cubes are not getting unloaded as in screenshot above. Except that they eventually do end up unloaded because vanilla unloads all cubes that are not loaded by any players every time it saves all chunks. But this is definitely not something that you should rely on. Maybe unload a cube when there is nothing left that depends on it?

**Comment by Barteks2x** [comment]
Another approach would be to keep them in unload queue and just skip them from unloading if there is something that depends on them. This way they will be automatically unloaded as soon as possible.

**Comment by Barteks2x** [comment]
It's not skipping them. It's removing them from unload queue and skipping.

**Comment by Cyclonit** [comment]
Does PlayerCubeMap not put them in the queeue again regardless?

**Comment by Barteks2x** [comment]
Nope. It unloads once. And then it assumes that it's unloaded.

**Comment by Barteks2x** [comment]
Also, why don't you just read the code?

**Comment by Cyclonit** [comment]
I apologize, I'm on the phone right now and commented faster than I was thinking.

**Comment by Cyclonit** [comment]
The ServerCubeCache no longer removes required cubes from the queue. But to be honest, I don't see a significant change.

File: src/main/java/cubicchunks/server/ServerCubeCache.java [hunk 25]
```diff
@@ -212,15 +214,16 @@ public boolean unloadQueuedChunks() {
 		final int MaxNumToUnload = 400;
 
 		// unload cubes
-		for (int i = 0; i < MaxNumToUnload && !this.cubesToUnload.isEmpty(); i++) {
-			long cubeAddress = this.cubesToUnload.poll();
+		Iterator<CubeCoords> iter = this.cubesToUnload.iterator();
+		while (iter.hasNext()) {
+			CubeCoords coords = iter.next();
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/server/ServerCubeCache.java]: you never limit the amount of unloaded cubes per tick. `unloadQueuedChunks` should be really called `tick`.

Comment (Cyclonit) [inline src/main/java/cubicchunks/server/ServerCubeCache.java]: You're right. But I assume limiting it would be a good idea. If you agree I'll add a simple counter and exit once the maximum number is reached. But I'll have to postpone doing so to tomorrow.

Comment (Barteks2x) [inline src/main/java/cubicchunks/server/ServerCubeCache.java]: This is how it was done before, and this is how vanilla does it.

File: src/main/java/cubicchunks/CustomCubicChunksWorldType.java [hunk 26]
```diff
@@ -44,8 +47,13 @@ public CustomCubicChunksWorldType() {
 		
 		// init the worldgen pipeline
 		GeneratorStage terrain = new IndependentGeneratorStage("terrain");
-		GeneratorStage features = new IndependentGeneratorStage("features");
-		GeneratorStage lighting = new IndependentGeneratorStage("lighting");
+
+		DependentGeneratorStage features = new DependentGeneratorStage("features", null);
+		features.setCubeDependency(new RegionDependency(features, new Vec3i(0, -1, 0), new Vec3i(0, 0, 0)));
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/CustomCubicChunksWorldType.java]: Change this to `IndependentGeneratorStage`, it doesn't require anything now.

**Comment by Barteks2x** [comment]
I know that unloadding still dowan't work correctly, but fixing that rewuored rewriting some parts of `ServerCubeCache` (changes from the branch I told you not to merge).
